File: backend/app/routers/portfolio.py
import logging
import os
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timedelta, timezone
from typing import Optional

import httpx
from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query, Request
from slowapi import Limiter
from slowapi.util import get_remote_address
from solders.pubkey import Pubkey
from sqlalchemy import desc, func, text
from sqlalchemy.orm import Session

from app.dependencies import get_db
from app.models.base import SessionLocal
from app.models.user_position import TrackedWallet, UserPosition, UserPositionEvent
from app.schemas import (
    TrackedWalletOut,
    UserPositionOut,
    UserPositionHistoryPoint,
    UserPositionEventOut,
    PositionsResponse,
    WalletStatusOut,
)
from app.services.kamino_position_fetcher import fetch_wallet_positions, fetch_wallet_events
from app.services.drift_position_fetcher import fetch_wallet_positions as fetch_drift_positions
from app.services.jupiter_position_fetcher import fetch_wallet_positions as fetch_jupiter_positions

logger = logging.getLogger(__name__)

router = APIRouter()
limiter = Limiter(key_func=get_remote_address)

HELIUS_API_KEY = os.getenv("HELIUS_API_KEY", "")


def _validate_wallet(wallet_address: str):
    try:
        Pubkey.from_string(wallet_address)
    except (ValueError, Exception):
        raise HTTPException(status_code=400, detail="Invalid Solana wallet address")


@router.get("/portfolio/{wallet_address}")
@limiter.limit("30/minute")
def get_portfolio(request: Request, wallet_address: str):
    """Read SPL token balances for a wallet via Helius RPC."""
    _validate_wallet(wallet_address)

    if not HELIUS_API_KEY:
        raise HTTPException(status_code=503, detail="Helius API key not configured")

    url = f"https://mainnet.helius-rpc.com/?api-key={HELIUS_API_KEY}"

    try:
        resp = httpx.post(
            url,
            json={
                "jsonrpc": "2.0",
                "id": 1,
                "method": "getTokenAccountsByOwner",
                "params": [
                    wallet_address,
                    {"programId": "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"},
                    {"encoding": "jsonParsed"},
                ],
            },
            timeout=15,
        )
        resp.raise_for_status()
        data = resp.json()
    except httpx.HTTPError as exc:
        raise HTTPException(status_code=502, detail=f"Helius RPC error: {exc}")

    if "error" in data:
        raise HTTPException(status_code=502, detail=data["error"].get("message", "RPC error"))

    accounts = data.get("result", {}).get("value", [])
    positions = []

    for account in accounts:
        info = account.get("account", {}).get("data", {}).get("parsed", {}).get("info", {})
        mint = info.get("mint", "")
        token_amount = info.get("tokenAmount", {})
        amount = token_amount.get("amount", "0")
        decimals = token_amount.get("decimals", 0)
        ui_amount = token_amount.get("uiAmount") or 0

        if ui_amount > 0:
            positions.append({
                "mint": mint,
                "symbol": None,
                "amount": float(amount),
                "decimals": decimals,
                "ui_amount": ui_amount,
            })

    return {"wallet": wallet_address, "positions": positions, "total_value_usd": 0}


# ---------------------------------------------------------------------------
# User Position Monitoring endpoints
# ---------------------------------------------------------------------------

def _latest_positions(db: Session, wallet_address: str, protocol: str | None = None, product_type: str | None = None):
    """Return the latest snapshot of UserPositions per protocol for a wallet."""
    latest_per_protocol = (
        db.query(
            UserPosition.protocol_slug,
            func.max(UserPosition.snapshot_at).label("latest_at"),
        )
        .filter(UserPosition.wallet_address == wallet_address)
        .group_by(UserPosition.protocol_slug)
        .subquery()
    )
    query = (
        db.query(UserPosition)
        .join(
            latest_per_protocol,
            (UserPosition.protocol_slug == latest_per_protocol.c.protocol_slug)
            & (UserPosition.snapshot_at == latest_per_protocol.c.latest_at)
            & (UserPosition.wallet_address == wallet_address),
        )
    )
    if protocol:
        query = query.filter(UserPosition.protocol_slug == protocol)
    if product_type:
        query = query.filter(UserPosition.product_type == product_type)
    return query.all()


def _store_positions(db: Session, positions: list[dict], now: datetime):
    """Store a list of position dicts as UserPosition rows."""
    from app.services.utils import store_position_rows
    store_position_rows(db, positions, now)


def _fetch_protocol(name: str, fetch_fn, wallet_address: str) -> list[dict] | None:
    db_local = SessionLocal()
    try:
        return fetch_fn(wallet_address, db_local)["positions"]
    except Exception as exc:
        logger.warning("%s fetch failed for %s: %s", name, wallet_address[:8], exc)
        return None
    finally:
        db_local.close()


def _background_fetch_and_store(wallet_address: str):
    """Background worker: fetch Kamino + Drift + Jupiter positions in parallel and store as snapshot."""
    db: Session = SessionLocal()
    try:
        tracked = (
            db.query(TrackedWallet)
            .filter(TrackedWallet.wallet_address == wallet_address)
            .first()
        )
        if not tracked:
            return
        tracked.fetch_status = "fetching"
        db.commit()

        now = datetime.now(timezone.utc)
        all_positions: list[dict] = []

        with ThreadPoolExecutor(max_workers=3) as executor:
            futures = [
                executor.submit(_fetch_protocol, "Kamino", fetch_wallet_positions, wallet_address),
                executor.submit(_fetch_protocol, "Drift", fetch_drift_positions, wallet_address),
                executor.submit(_fetch_protocol, "Jupiter", fetch_jupiter_positions, wallet_address),
            ]
            for future in as_completed(futures):
                result = future.result()
                if result is not None:
                    all_positions.extend(result)

        if not all_positions:
            logger.warning("All fetches failed for %s — keeping old snapshot", wallet_address[:8])
            tracked.fetch_status = "ready"
            db.commit()
            return

        _store_positions(db, all_positions, now)

        tracked.last_fetched_at = now
        tracked.fetch_status = "ready"
        db.commit()
        logger.info("Background fetch complete for %s: %d positions", wallet_address[:8], len(all_positions))

    except Exception as exc:
        logger.error("Background fetch failed for %s: %s", wallet_address[:8], exc)
        db.rollback()
        try:
            tracked = (
                db.query(TrackedWallet)
                .filter(TrackedWallet.wallet_address == wallet_address)
                .first()
            )
            if tracked:
                tracked.fetch_status = "error"
                db.commit()
        except Exception:
            db.rollback()
    finally:
        db.close()


@router.post("/portfolio/{wallet_address}/track")
@limiter.limit("5/minute")
def track_wallet(
    request: Request,
    wallet_address: str,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
):
    """Register a wallet for tracking. Returns immediately; fetch runs in background."""
    _validate_wallet(wallet_address)

    # Upsert TrackedWallet
    wallet = (
        db.query(TrackedWallet)
        .filter(TrackedWallet.wallet_address == wallet_address)
        .first()
    )
    if wallet:
        wallet.is_active = True
    else:
        wallet = TrackedWallet(wallet_address=wallet_address)
        db.add(wallet)
    db.flush()

    # Check if we already have positions from a prior snapshot
    has_positions = (
        db.query(UserPosition.id)
        .filter(UserPosition.wallet_address == wallet_address)
        .first()
    )
    if has_positions:
        # Return cached positions (latest per protocol), refresh in background
        positions = _latest_positions(db, wallet_address)
        position_dicts = [
            {
                "wallet_address": p.wallet_address,
                "protocol_slug": p.protocol_slug,
                "product_type": p.product_type,
                "external_id": p.external_id,
                "opportunity_id": p.opportunity_id,
                "deposit_amount": float(p.deposit_amount) if p.deposit_amount else None,
                "deposit_amount_usd": float(p.deposit_amount_usd) if p.deposit_amount_usd else None,
                "pnl_usd": float(p.pnl_usd) if p.pnl_usd else None,
                "pnl_pct": float(p.pnl_pct) if p.pnl_pct else None,
                "initial_deposit_usd": float(p.initial_deposit_usd) if p.initial_deposit_usd else None,
                "opened_at": p.opened_at.isoformat() if p.opened_at else None,
                "held_days": float(p.held_days) if p.held_days else None,
                "apy": float(p.apy) if p.apy else None,
                "is_closed": p.is_closed,
                "closed_at": p.closed_at.isoformat() if p.closed_at else None,
                "close_value_usd": float(p.close_value_usd) if p.close_value_usd else None,
                "token_symbol": p.token_symbol,
                "extra_data": p.extra_data,
            }
            for p in positions
        ]
        total_value = sum(float(p.deposit_amount_usd or 0) for p in positions)
        total_pnl = sum(float(p.pnl_usd or 0) for p in positions if p.pnl_usd is not None)

        wallet.fetch_status = "fetching"
        db.commit()
        background_tasks.add_task(_background_fetch_and_store, wallet_address)

        return {
            "wallet": wallet_address,
            "positions": position_dicts,
            "summary": {
                "total_value_usd": total_value,
                "total_pnl_usd": total_pnl,
                "position_count": len(position_dicts),
            },
            "fetch_status": "ready",
        }

    # No prior data — kick off background fetch, return empty
    wallet.fetch_status = "fetching"
    db.commit()
    background_tasks.add_task(_background_fetch_and_store, wallet_address)

    return {
        "wallet": wallet_address,
        "positions": [],
        "summary": {
            "total_value_usd": 0,
            "total_pnl_usd": 0,
            "position_count": 0,
        },
        "fetch_status": "fetching",
    }


@router.get("/portfolio/{wallet_address}/status", response_model=WalletStatusOut)
@limiter.limit("60/minute")
def get_wallet_status(request: Request, wallet_address: str, db: Session = Depends(get_db)):
    """Return fetch status for a tracked wallet (for frontend polling)."""
    _validate_wallet(wallet_address)
    tracked = (
        db.query(TrackedWallet)
        .filter(TrackedWallet.wallet_address == wallet_address)
        .first()
    )
    if not tracked:
        raise HTTPException(status_code=404, detail="Wallet not tracked")
    return tracked


@router.get("/portfolio/{wallet_address}/positions", response_model=list[UserPositionOut])
@limiter.limit("60/minute")
def get_positions(
    request: Request,
    wallet_address: str,
    protocol: Optional[str] = Query(None),
    product_type: Optional[str] = Query(None),
    db: Session = Depends(get_db),
):
    """Return the latest snapshot of positions from DB."""
    _validate_wallet(wallet_address)
    return _latest_positions(db, wallet_address, protocol, product_type)


@router.get("/portfolio/{wallet_address}/positions/history")
@limiter.limit("60/minute")
def get_position_history(
    request: Request,
    wallet_address: str,
    period: str = Query("7d", pattern="^(7d|30d|90d)$"),
    external_id: Optional[str] = Query(None),
    limit: int = Query(500, ge=1, le=2000),
    offset: int = Query(0, ge=0),
    db: Session = Depends(get_db),
):
    """Return time-series position snapshots for PnL charts."""
    _validate_wallet(wallet_address)

    days = {"7d": 7, "30d": 30, "90d": 90}[period]
    cutoff = datetime.now(timezone.utc) - timedelta(days=days)

    # Per-position history (single position, low row count)
    if external_id:
        rows = (
            db.query(UserPosition)
            .filter(
                UserPosition.wallet_address == wallet_address,
                UserPosition.snapshot_at >= cutoff,
                UserPosition.external_id == external_id,
            )
            .order_by(UserPosition.snapshot_at)
            .offset(offset)
            .limit(limit)
            .all()
        )
        return [
            UserPositionHistoryPoint(
                snapshot_at=r.snapshot_at,
                deposit_amount_usd=float(r.deposit_amount_usd) if r.deposit_amount_usd else None,
                pnl_usd=float(r.pnl_usd) if r.pnl_usd else None,
                pnl_pct=float(r.pnl_pct) if r.pnl_pct else None,
            )
            for r in rows
        ]

    # Aggregate view — bucket by time interval to keep chart data small
    bucket_interval = {"7d": "1 hour", "30d": "4 hours", "90d": "12 hours"}[period]

    # Step 1: sum all positions per snapshot timestamp
    per_snapshot = (
        db.query(
            UserPosition.snapshot_at,
            func.sum(UserPosition.deposit_amount_usd).label("total_usd"),
            func.sum(UserPosition.pnl_usd).label("total_pnl"),
        )
        .filter(
            UserPosition.wallet_address == wallet_address,
            UserPosition.snapshot_at >= cutoff,
        )
        .group_by(UserPosition.snapshot_at)
        .subquery()
    )

    # Step 2: pick the latest snapshot per time bucket via DISTINCT ON
    bucket_expr = func.date_bin(
        text(f"'{bucket_interval}'"), per_snapshot.c.snapshot_at, cutoff,
    )

    rows = (
        db.query(
            bucket_expr.label("bucket"),
            per_snapshot.c.total_usd.label("deposit_amount_usd"),
            per_snapshot.c.total_pnl.label("pnl_usd"),
        )
        .distinct(bucket_expr)
        .order_by(bucket_expr, per_snapshot.c.snapshot_at.desc())
        .all()
    )

    return [
        UserPositionHistoryPoint(
            snapshot_at=row.bucket,
            deposit_amount_usd=float(row.deposit_amount_usd) if row.deposit_amount_usd else 0.0,
            pnl_usd=float(row.pnl_usd) if row.pnl_usd else 0.0,
            pnl_pct=None,
        )
        for row in rows
    ]


@router.get("/portfolio/{wallet_address}/events", response_model=list[UserPositionEventOut])
@limiter.limit("60/minute")
def get_events(
    request: Request,
    wallet_address: str,
    protocol: Optional[str] = Query(None),
    product_type: Optional[str] = Query(None),
    limit: int = Query(50, le=200),
    db: Session = Depends(get_db),
):
    """Return deposit/withdraw event history."""
    _validate_wallet(wallet_address)

    query = db.query(UserPositionEvent).filter(
        UserPositionEvent.wallet_address == wallet_address,
    )
    if protocol:
        query = query.filter(UserPositionEvent.protocol_slug == protocol)
    if product_type:
        query = query.filter(UserPositionEvent.product_type == product_type)

    return query.order_by(desc(UserPositionEvent.event_at)).limit(limit).all()

File: backend/src/monitor/routes/portfolio.ts
/**
 * Portfolio routes — port of backend/app/routers/portfolio.py
 */
import type { FastifyInstance } from "fastify";
import { sql, eq, and, gte, desc, asc } from "drizzle-orm";
import { db } from "../db/connection.js";
import { trackedWallets, userPositions, userPositionEvents } from "../db/schema.js";
import { NotFoundError } from "../../shared/error-handler.js";
import { postJson } from "../../shared/http.js";
import { logger } from "../../shared/logger.js";
import { getSymbolForMint, isStablecoinMint } from "../../shared/constants.js";
import { validateWallet, storePositionRows, storeEventsBatch } from "../services/utils.js";
import { fetchWalletPositions as fetchKaminoPositions } from "../services/kamino-position-fetcher.js";
import { fetchWalletPositions as fetchDriftPositions } from "../services/drift-position-fetcher.js";
import { fetchWalletPositions as fetchJupiterPositions } from "../services/jupiter-position-fetcher.js";
import { fetchWalletPositions as fetchExponentPositions } from "../services/exponent-position-fetcher.js";
import { numOrNull } from "../../shared/utils.js";
import {
  PositionsQuery,
  PositionHistoryQuery,
  EventsQuery,
  PortfolioAnalyticsOut,
  SyncBody,
} from "./schemas.js";
import { monitorService } from "../service.js";

// ---------------------------------------------------------------------------
// Background fetch (fire-and-forget)
// ---------------------------------------------------------------------------

const FETCH_TIMEOUT_MS = 5 * 60 * 1000; // 5 min — defense against hung protocol APIs
const inFlightWallets = new Set<string>();

async function backgroundFetchAndStore(walletAddress: string) {
  if (inFlightWallets.has(walletAddress)) {
    logger.debug({ wallet: walletAddress.slice(0, 8) }, "Background fetch already in flight — skipping");
    return;
  }
  inFlightWallets.add(walletAddress);
  try {
    // Mark as fetching
    await db
      .update(trackedWallets)
      .set({ fetch_status: "fetching" })
      .where(eq(trackedWallets.wallet_address, walletAddress));

    const now = new Date();

    // Fetch all protocols in parallel, with a hard timeout
    const [kaminoResult, driftResult, jupiterResult, exponentResult] =
      await Promise.race([
        Promise.allSettled([
          fetchKaminoPositions(walletAddress),
          fetchDriftPositions(walletAddress),
          fetchJupiterPositions(walletAddress, db),
          fetchExponentPositions(walletAddress),
        ]),
        new Promise<never>((_, reject) =>
          setTimeout(() => reject(new Error("Background fetch timeout (5min)")), FETCH_TIMEOUT_MS),
        ),
      ]);

    const allPositions = [
      ...(kaminoResult.status === "fulfilled"
        ? kaminoResult.value.positions
        : []),
      ...(driftResult.status === "fulfilled"
        ? driftResult.value.positions
        : []),
      ...(jupiterResult.status === "fulfilled"
        ? jupiterResult.value.positions
        : []),
      ...(exponentResult.status === "fulfilled"
        ? exponentResult.value.positions
        : []),
    ];

    // Also collect events from Drift IF
    const driftEvents =
      driftResult.status === "fulfilled"
        ? driftResult.value.events
        : [];

    if (allPositions.length === 0) {
      logger.warn(
        { wallet: walletAddress.slice(0, 8) },
        "All fetches failed — keeping old snapshot",
      );
      await db
        .update(trackedWallets)
        .set({ fetch_status: "ready" })
        .where(eq(trackedWallets.wallet_address, walletAddress));
      return;
    }

    await storePositionRows(db, allPositions, now);
    if (driftEvents.length > 0) {
      await storeEventsBatch(db, driftEvents);
    }

    await db
      .update(trackedWallets)
      .set({ last_fetched_at: now, fetch_status: "ready" })
      .where(eq(trackedWallets.wallet_address, walletAddress));

    logger.info(
      { wallet: walletAddress.slice(0, 8), count: allPositions.length },
      "Background fetch complete",
    );
  } catch (err) {
    logger.error(
      { err, wallet: walletAddress.slice(0, 8) },
      "Background fetch failed",
    );
    try {
      await db
        .update(trackedWallets)
        .set({ fetch_status: "error" })
        .where(eq(trackedWallets.wallet_address, walletAddress));
    } catch {
      // ignore
    }
  } finally {
    inFlightWallets.delete(walletAddress);
  }
}

// ---------------------------------------------------------------------------
// Routes
// ---------------------------------------------------------------------------

export async function portfolioRoutes(app: FastifyInstance) {
  // -----------------------------------------------------------------------
  // GET /portfolio/:wallet — SPL token balances
  // -----------------------------------------------------------------------
  app.get<{ Params: { wallet: string } }>("/portfolio/:wallet", {
    config: { rateLimit: { max: 30, timeWindow: "1 minute" } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);

      const heliusApiKey = process.env.HELIUS_API_KEY;
      if (!heliusApiKey) {
        const err = new Error("Helius API key not configured") as Error & { statusCode: number };
        err.statusCode = 503;
        throw err;
      }

      const url = `https://mainnet.helius-rpc.com/?api-key=${heliusApiKey}`;

      // Fetch from both Token Program and Token-2022 in parallel
      const programIds = [
        "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",   // Token Program
        "TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb",   // Token-2022
      ];

      let allAccounts: Record<string, unknown>[] = [];
      try {
        const results = await Promise.all(
          programIds.map((programId) =>
            postJson(url, {
              jsonrpc: "2.0",
              id: 1,
              method: "getTokenAccountsByOwner",
              params: [wallet, { programId }, { encoding: "jsonParsed" }],
            }) as Promise<Record<string, unknown>>
          ),
        );

        for (const data of results) {
          if (data.error) {
            const e = new Error(
              ((data.error as Record<string, unknown>).message as string) ?? "RPC error",
            ) as Error & { statusCode: number };
            e.statusCode = 502;
            throw e;
          }
          const accounts =
            ((data.result as Record<string, unknown>)?.value as Record<string, unknown>[]) ?? [];
          allAccounts = allAccounts.concat(accounts);
        }
      } catch (err) {
        if ((err as Error & { statusCode?: number }).statusCode) throw err;
        const e = new Error(`Helius RPC error`) as Error & { statusCode: number };
        e.statusCode = 502;
        throw e;
      }

      const positions = [];
      for (const account of allAccounts) {
        const info =
          (
            (
              (account.account as Record<string, unknown>)
                ?.data as Record<string, unknown>
            )?.parsed as Record<string, unknown>
          )?.info as Record<string, unknown> | undefined;
        if (!info) continue;

        const mint = (info.mint as string) ?? "";
        const tokenAmount =
          (info.tokenAmount as Record<string, unknown>) ?? {};
        const amount = (tokenAmount.amount as string) ?? "0";
        const decimals = Number(tokenAmount.decimals ?? 0);
        const uiAmount = Number(tokenAmount.uiAmount ?? 0);

        if (uiAmount > 0) {
          positions.push({
            mint,
            symbol: getSymbolForMint(mint),
            amount: Number(amount),
            decimals,
            ui_amount: uiAmount,
            is_stablecoin: isStablecoinMint(mint),
          });
        }
      }

      const totalStablecoinUsd = positions
        .filter((p) => p.is_stablecoin)
        .reduce((sum, p) => sum + p.ui_amount, 0);

      return { wallet, positions, total_value_usd: totalStablecoinUsd };
    },
  });

  // -----------------------------------------------------------------------
  // POST /portfolio/:wallet/track — register wallet, background fetch
  // -----------------------------------------------------------------------
  app.post<{ Params: { wallet: string } }>("/portfolio/:wallet/track", {
    config: { rateLimit: { max: 5, timeWindow: "1 minute" } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);

      // Upsert TrackedWallet
      const existing = await db
        .select()
        .from(trackedWallets)
        .where(eq(trackedWallets.wallet_address, wallet))
        .limit(1);

      if (existing.length > 0) {
        await db
          .update(trackedWallets)
          .set({ is_active: true })
          .where(eq(trackedWallets.wallet_address, wallet));
      } else {
        await db
          .insert(trackedWallets)
          .values({ wallet_address: wallet });
      }

      // Check for cached positions
      const hasPositions = await db
        .select({ id: userPositions.id })
        .from(userPositions)
        .where(eq(userPositions.wallet_address, wallet))
        .limit(1);

      if (hasPositions.length > 0) {
        const result = await monitorService.getPortfolioPositions(wallet);

        // Update status and kick off background refresh
        await db
          .update(trackedWallets)
          .set({ fetch_status: "fetching" })
          .where(eq(trackedWallets.wallet_address, wallet));

        // Fire-and-forget background fetch
        void backgroundFetchAndStore(wallet);

        return {
          ...result,
          fetch_status: "fetching",
        };
      }

      // No prior data — start background fetch
      await db
        .update(trackedWallets)
        .set({ fetch_status: "fetching" })
        .where(eq(trackedWallets.wallet_address, wallet));

      void backgroundFetchAndStore(wallet);

      return {
        wallet,
        positions: [],
        summary: {
          total_value_usd: 0,
          total_pnl_usd: 0,
          position_count: 0,
        },
        fetch_status: "fetching",
      };
    },
  });

  // -----------------------------------------------------------------------
  // POST /portfolio/:wallet/sync — sync a single position after transaction
  // -----------------------------------------------------------------------
  app.post<{ Params: { wallet: string } }>("/portfolio/:wallet/sync", {
    config: { rateLimit: { max: 10, timeWindow: "1 minute" } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);

      const body = SyncBody.parse(request.body);
      await monitorService.syncPosition(wallet, body.opportunity_id, body.metadata);
      return { ok: true };
    },
  });

  // -----------------------------------------------------------------------
  // GET /portfolio/:wallet/status
  // -----------------------------------------------------------------------
  app.get<{ Params: { wallet: string } }>("/portfolio/:wallet/status", {
    config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);

      const rows = await db
        .select()
        .from(trackedWallets)
        .where(eq(trackedWallets.wallet_address, wallet))
        .limit(1);

      if (rows.length === 0) throw new NotFoundError("Wallet not tracked");

      return {
        wallet_address: rows[0].wallet_address,
        fetch_status: rows[0].fetch_status,
        last_fetched_at: rows[0].last_fetched_at,
      };
    },
  });

  // -----------------------------------------------------------------------
  // GET /portfolio/:wallet/analytics
  // -----------------------------------------------------------------------
  app.get<{ Params: { wallet: string } }>("/portfolio/:wallet/analytics", {
    config: { rateLimit: { max: 30, timeWindow: "1 minute" } },
    schema: { response: { 200: PortfolioAnalyticsOut } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);
      return monitorService.getPortfolioAnalytics(wallet);
    },
  });

  // -----------------------------------------------------------------------
  // GET /portfolio/:wallet/positions
  // -----------------------------------------------------------------------
  app.get<{ Params: { wallet: string } }>("/portfolio/:wallet/positions", {
    config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);
      const q = PositionsQuery.parse(request.query);

      const result = await monitorService.getPortfolioPositions(
        wallet,
        q.protocol,
        q.product_type,
        q.asset_class,
      );
      return result.positions;
    },
  });

  // -----------------------------------------------------------------------
  // GET /portfolio/:wallet/positions/history
  // -----------------------------------------------------------------------
  app.get<{ Params: { wallet: string } }>(
    "/portfolio/:wallet/positions/history",
    {
      config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
      handler: async (request) => {
        const { wallet } = request.params;
        validateWallet(wallet);
        const q = PositionHistoryQuery.parse(request.query);

        try {
          const days = { "7d": 7, "30d": 30, "90d": 90 }[q.period] ?? 7;
          const cutoff = new Date(Date.now() - days * 86_400_000);

          // Per-position history
          if (q.external_id) {
            const rows = await db
              .select()
              .from(userPositions)
              .where(
                and(
                  eq(userPositions.wallet_address, wallet),
                  gte(userPositions.snapshot_at, cutoff),
                  eq(userPositions.external_id, q.external_id),
                ),
              )
              .orderBy(asc(userPositions.snapshot_at))
              .offset(q.offset)
              .limit(q.limit);

            return rows.map((r) => ({
              snapshot_at: r.snapshot_at,
              deposit_amount_usd: numOrNull(r.deposit_amount_usd),
              pnl_usd: numOrNull(r.pnl_usd),
              pnl_pct: numOrNull(r.pnl_pct),
            }));
          }

          // Aggregate history with time bucketing
          const bucketInterval = {
            "7d": "4 hours",
            "30d": "8 hours",
            "90d": "12 hours",
          }[q.period] ?? "4 hours";

          const rows = await db.execute(sql`
            WITH per_snapshot AS (
              SELECT
                snapshot_at,
                SUM(deposit_amount_usd::numeric) as total_usd,
                SUM(pnl_usd::numeric) as total_pnl
              FROM monitor.user_positions
              WHERE wallet_address = ${wallet}
                AND snapshot_at >= ${cutoff}
              GROUP BY snapshot_at
            ),
            bucketed AS (
              SELECT
                date_bin(${sql.raw(`'${bucketInterval}'`)}, snapshot_at, ${cutoff}::timestamp) as bucket,
                total_usd,
                total_pnl,
                snapshot_at
              FROM per_snapshot
            )
            SELECT DISTINCT ON (bucket)
              bucket,
              total_usd as deposit_amount_usd,
              total_pnl as pnl_usd
            FROM bucketed
            ORDER BY bucket, snapshot_at DESC
          `);

          return rows.rows.map((r: Record<string, unknown>) => ({
            snapshot_at: r.bucket as Date,
            deposit_amount_usd: r.deposit_amount_usd
              ? Number(r.deposit_amount_usd)
              : 0,
            pnl_usd: r.pnl_usd ? Number(r.pnl_usd) : 0,
            pnl_pct: null,
          }));
        } catch (err) {
          logger.error({ err, wallet: wallet.slice(0, 8) }, "Position history query failed");
          throw err;
        }
      },
    },
  );

  // -----------------------------------------------------------------------
  // GET /portfolio/:wallet/events
  // -----------------------------------------------------------------------
  app.get<{ Params: { wallet: string } }>("/portfolio/:wallet/events", {
    config: { rateLimit: { max: 60, timeWindow: "1 minute" } },
    handler: async (request) => {
      const { wallet } = request.params;
      validateWallet(wallet);
      const q = EventsQuery.parse(request.query);

      const conditions = [
        eq(userPositionEvents.wallet_address, wallet),
      ];
      if (q.protocol)
        conditions.push(eq(userPositionEvents.protocol_slug, q.protocol));
      if (q.product_type)
        conditions.push(
          eq(userPositionEvents.product_type, q.product_type),
        );

      const rows = await db
        .select()
        .from(userPositionEvents)
        .where(and(...conditions))
        .orderBy(desc(userPositionEvents.event_at))
        .limit(q.limit);

      return rows.map((r) => ({
        id: r.id,
        wallet_address: r.wallet_address,
        protocol_slug: r.protocol_slug,
        product_type: r.product_type,
        external_id: r.external_id,
        event_type: r.event_type,
        amount: numOrNull(r.amount),
        amount_usd: numOrNull(r.amount_usd),
        tx_signature: r.tx_signature,
        event_at: r.event_at,
        extra_data: r.extra_data,
      }));
    },
  });
}
